WcaOnRails/app/models/upload_json.rb

```
@@ -0,0 +1,136 @@
+                new_scramble_attributes = {
+                  competitionId: competition_id,
+                  eventId: event["eventId"],
+                  roundTypeId: round["roundId"],
+                  groupId: group["group"],
+                  isExtra: scramble_type == "extraScrambles",
+                  scrambleNum: index+1,
+                  scramble: scramble,
+                }
+                scrambles_to_import << Scramble.new(new_scramble_attributes)
+              end
+            end
+          end
+        end
+      end
+      begin
+        ActiveRecord::Base.transaction do
+          InboxPerson.where(competitionId: competition_id).delete_all
+          InboxResult.where(competitionId: competition_id).delete_all
+          Scramble.where(competitionId: competition_id).delete_all
+          InboxPerson.import!(persons_to_import)
+          Scramble.import!(scrambles_to_import)
+          InboxResult.import!(results_to_import)
+        end
+        true
+      rescue ActiveRecord::RecordNotUnique
+        errors.add(:results_file, "Duplicate personId in JSON.")
+        false
+      rescue ActiveRecord::RecordInvalid => invalid
+        object = invalid.record
+        if object.class == Scramble
+          errors.add(:results_file, "Scramble in '#{Round.name_from_attributes_id(object.eventId, object.roundTypeId)}' is invalid (#{invalid.message}), please fix it!")
+        elsif object.class == InboxPerson
+          errors.add(:results_file, "Person #{object.name} is invalid (#{invalid.message}), please fix it!")
+        elsif object.class == InboxResult
+          errors.add(:results_file, "Result for person #{object.personId} in '#{Round.name_from_attributes_id(object.eventId, object.roundTypeId)}' is invalid (#{invalid.message}), please fix it!")
+        else
+          # FIXME: that's actually not supposed to happen, as the only 3 types of records we create are above
```

If this is not supposed to happen, but (hypothetically) we don't want it to fail silently, then we could just call `raise` here (without arguments it just raises the last exception). A validation error is also alright, it shouldn't happen afterall ^^

---

I'm afraid just an exception would result in the standard "something went wrong" page, and I think it would be more helpful to have an error in the form.